packages/contracts/contracts/RollupMerkleUtils.sol

```diff
@@ -0,0 +1,268 @@
+                nextLevelLength += 1;
+            }
+        }
+
+        // Alright! We should be left with a single node! Return it...
+        return nodes[0];
+    }
+
+    /**
+     * @notice Verify an inclusion proof.
+     * @param _root The root of the tree we are verifying inclusion for.
+     * @param _dataBlock The data block we're verifying inclusion for.
+     * @param _path The path from the leaf to the root.
+     * @param _siblings The sibling nodes along the way.
+     * @return The next level of the tree
+     */
+    function verify(bytes32 _root, bytes memory _dataBlock, uint _path, bytes32[] memory _siblings) public pure returns (bool) {
+        // First compute the leaf node
+        bytes32 computedNode = keccak256(_dataBlock);
+        for (uint i = 0; i < _siblings.length; i++) {
+            bytes32 sibling = _siblings[i];
+            uint8 isComputedRightSibling = getNthBitFromRight(_path, i);
+            if (isComputedRightSibling == 0) {
+                computedNode = getParent(computedNode, sibling);
+            } else {
+                computedNode = getParent(sibling, computedNode);
+            }
+        }
+        // Check if the computed node (_root) is equal to the provided root
+        return computedNode == _root;
+    }
+
+    /**
+     * @notice Update the stored tree / root with a particular dataBlock at some path (no siblings needed)
+     * @param _dataBlock The data block we're storing/verifying
+     * @param _path The path from the leaf to the root / the index of the leaf.
+     */
+    function update(bytes memory _dataBlock, uint _path) public {
+        bytes32[] memory siblings = getSiblings(_path);
+        store(_dataBlock, _path, siblings);
+    }
+
+    /**
+     * @notice Store a particular merkle proof & verify that the root did not change.
+     * @param _dataBlock The data block we're storing/verifying
+     * @param _path The path from the leaf to the root / the index of the leaf.
+     * @param _siblings The sibling nodes along the way.
+     */
+    function storeAndVerify(bytes memory _dataBlock, uint _path, bytes32[] memory _siblings) public {
+        bytes32 oldRoot = tree.root;
+        store(_dataBlock, _path, _siblings);
+        require(tree.root == oldRoot, "Failed same root verification check! This was an inclusion proof for a different tree!");
+    }
+
+
+    /**
+     * @notice Store a particular leaf & intermediate nodes in the tree
+     * @param _dataBlock The data block we're storing.
+     * @param _path The path from the leaf to the root / the index of the leaf.
+     * @param _siblings The sibling nodes along the way.
+     */
+    function store(bytes memory _dataBlock, uint _path, bytes32[] memory _siblings) public {
+        // First compute the leaf node
+        bytes32 computedNode = keccak256(_dataBlock);
+        for (uint i = 0; i < _siblings.length; i++) {
+            bytes32 parent;
+            bytes32 sibling = _siblings[i];
```

Comment:
*Important* I think our siblings are in different orders. For me the last sibling is the leaf sibling.
```
siblings = [
root_child_sibling
...
leaf_sibling
]
```

 If this makes the most sense, I'll update. Let me know.